vnext/JSI/Shared/ChakraApi.cpp

```diff
@@ -0,0 +1,439 @@
+    errorString << "A call to Chakra API returned error code 0x" << std::hex << errorCode << '.';
+    throw std::exception(errorString.str().c_str());
+  }
+}
+
+void ChakraApi::ThrowNativeException(char const *errorMessage) {
+  if (auto thrower = ExceptionThrowerHolder::Get()) {
+    thrower->ThrowNativeExceptionOverride(errorMessage);
+  } else {
+    throw std::exception(errorMessage);
+  }
+}
+
+uint32_t ChakraApi::AddRef(JsRef ref) {
+  uint32_t result;
+  ChakraVerifyJsErrorElseThrow(JsAddRef(ref, &result));
+  return result;
+}
+
+uint32_t ChakraApi::Release(JsRef ref) {
+  uint32_t result;
+  ChakraVerifyJsErrorElseThrow(JsRelease(ref, &result));
+  return result;
+}
+
+JsContextRef ChakraApi::CreateContext(JsRuntimeHandle runtime) {
+  JsContextRef context{JS_INVALID_REFERENCE};
+  ChakraVerifyJsErrorElseThrow(JsCreateContext(runtime, &context));
+  return context;
+}
+
+JsContextRef ChakraApi::GetCurrentContext() {
+  JsContextRef context{JS_INVALID_REFERENCE};
+  ChakraVerifyJsErrorElseThrow(JsGetCurrentContext(&context));
+  return context;
+}
+
+void ChakraApi::SetCurrentContext(JsContextRef context) {
+  ChakraVerifyJsErrorElseThrow(JsSetCurrentContext(context));
+}
+
+JsPropertyIdRef ChakraApi::GetPropertyIdFromName(wchar_t const *name) {
+  JsPropertyIdRef propertyId;
+  ChakraVerifyJsErrorElseThrow(JsGetPropertyIdFromName(name, &propertyId));
+  return propertyId;
+}
+
+JsPropertyIdRef ChakraApi::GetPropertyIdFromString(JsValueRef value) {
+  return GetPropertyIdFromName(StringToPointer(value).data());
+}
+
+JsPropertyIdRef ChakraApi::GetPropertyIdFromName(std::string_view name) {
+  ChakraVerifyElseThrow(name.data(), "Property name cannot be a nullptr.");
+
+  JsPropertyIdRef propertyId{JS_INVALID_REFERENCE};
+  // We use a #ifdef here because we can avoid a UTF-8 to UTF-16 conversion
+  // using ChakraCore's JsCreatePropertyId API.
+#ifdef CHAKRACORE
+  ChakraVerifyJsErrorElseThrow(JsCreatePropertyId(name.data(), name.length(), &propertyId));
+#else
+  std::wstring utf16 = Common::Unicode::Utf8ToUtf16(name.data(), name.length());
+  ChakraVerifyJsErrorElseThrow(JsGetPropertyIdFromName(utf16.data(), &propertyId));
+#endif
+  return propertyId;
+}
+
+wchar_t const *ChakraApi::GetPropertyNameFromId(JsPropertyIdRef propertyId) {
+  ChakraVerifyElseThrow(
+      GetPropertyIdType(propertyId) == JsPropertyIdTypeString,
+      "It is illegal to retrieve the name associated with a property symbol.");
+
+  wchar_t const *name;
+  ChakraVerifyJsErrorElseThrow(JsGetPropertyNameFromId(propertyId, &name));
+  return name;
+}
+
+JsValueRef ChakraApi::GetPropertyStringFromId(JsPropertyIdRef propertyId) {
+  return PointerToString(GetPropertyNameFromId(propertyId));
+}
+
+JsValueRef ChakraApi::GetSymbolFromPropertyId(JsPropertyIdRef propertyId) {
+  ChakraVerifyElseThrow(
+      GetPropertyIdType(propertyId) == JsPropertyIdTypeSymbol,
+      "It is illegal to retrieve the symbol associated with a property name.");
+
```

Comment:
consider explaining why instead of what (same in 138)

Reply:
The error messages already explain 'why', not 'what'. The JsPropertyIdRef could be based on a string or on a symbol. Thus, some operation are only permitted on one of the types. Feel free to offer an alternative error text.

---
In reply to: [470463462](https://github.com/microsoft/react-native-windows/pull/5710#discussion_r470463462) [](ancestors = 470463462)